src/exchange.ts

```diff
@@ -0,0 +1,235 @@
+    pair: SwapPairInterface,
+    amount: number,
+    price?: number,
+    tags: DecodedTag[] = []
+  ): Promise<string> {
+    // Validate hashes
+    if (
+      !this.utils.validateHash(pair.from) ||
+      !this.utils.validateHash(pair.to)
+    )
+      throw new Error(
+        "Invalid ID in pair. Must be a valid SmartWeave contract ID"
+      );
+
+    const contract = this.smartweave
+      .contract(this.utils.CLOB_CONTRACT)
+      .connect(this.wallet);
+
+    const orderID = await cacheContractHook(async () => {
+      // Transfer input tokens to the orderbook
+      const transferID = await this.token.transfer(
+        amount,
+        pair.from,
+        this.utils.CLOB_CONTRACT,
+        [{ name: "Type", value: "Send-Input" }]
+      );
+
+      // Create the swap interaction
+      return await contract.writeInteraction(
+        {
+          function: "createOrder",
+          transaction: transferID,
+          pair: [pair.from, pair.to],
+          price: price,
+        },
+        [
+          {
+            name: "Exchange",
+            value: "Verto",
+          },
+          {
+            name: "Action",
+            value: "Order",
+          },
+          ...tags,
+        ]
+      );
+    }, [pair.from, this.utils.CLOB_CONTRACT, pair.to]);
+
+    if (orderID === null) throw new Error("Could not create order");
+
+    // Create exchange fee
+    await this.utils.createFee(amount, pair.from, orderID, "exchange");
+
+    // Create VRT holder fee
+    await this.utils.createFee(amount, pair.from, orderID, "token_holder");
+
+    // Call Discord hook
+    axios.post(`https://hook.verto.exchange/api/transaction?id=${orderID}`);
+
+    return orderID;
+  }
+
+  /**
+   * Cancel an order. This will return the non-filled tokens for an order and
+   * remove it from the orderbook
+   * @param orderID The transaction id of the swap
+   * @returns The transaction id of the cancel
+   */
+  async cancel(orderID: string): Promise<string> {
+    const contract = this.smartweave
+      .contract(this.utils.CLOB_CONTRACT)
+      .connect(this.wallet);
+
+    const transactionID = await cacheContractHook(
+      async () =>
+        contract.writeInteraction(
+          {
+            function: "cancelOrder",
+            orderID,
+          },
+          [
+            {
+              name: "Exchange",
+              value: "Verto",
+            },
+            {
+              name: "Action",
+              value: "CancelOrder",
+            },
+          ]
+        ),
+      this.utils.CLOB_CONTRACT
+    );
+
+    if (!transactionID) throw new Error("Order could not be cancelled");
+
+    return transactionID;
+  }
+
+  /**
+   * Fetches the order book for a specific token from the CLOB contract
+   * @param input Token contract ID or token pair
+   * @returns List of orders
+   */
+  async getOrderBook(input: string | TokenPair): Promise<OrderInterface[]> {
+    // get clob contract state
+    const clobContractState: {
```

Comment:
Isolate the type inside an interface, so it can be accessible if anything.

Reply:
What type?